crates/swc_css_ast/src/token.rs

```diff
@@ -35,135 +35,202 @@ pub enum NumberType {
-        value: JsWord,
-        raw_value: Atom,
-    },
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, EqIgnoreSpan)]
+#[cfg_attr(
+    feature = "rkyv",
+    derive(rkyv::Archive, rkyv::Serialize, rkyv::Deserialize)
+)]
+pub struct BadUrlToken {
+    /// Name and value
+    pub raw: Box<(Atom, Atom)>,
+}
 
-    BadUrl {
-        #[cfg_attr(feature = "rkyv", with(swc_atoms::EncodeJsWord))]
-        name: JsWord,
-        raw_name: Atom,
-        raw_value: Atom,
-    },
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, EqIgnoreSpan)]
+#[cfg_attr(
+    feature = "rkyv",
+    derive(rkyv::Archive, rkyv::Serialize, rkyv::Deserialize)
+)]
+pub struct DelimToken {
+    pub value: char,
+}
 
-    Delim {
-        value: char,
-    },
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, EqIgnoreSpan)]
+#[cfg_attr(
+    feature = "rkyv",
+    derive(rkyv::Archive, rkyv::Serialize, rkyv::Deserialize)
+)]
+pub struct NumberToken {
+    pub value: f64,
+    pub raw: Atom,
+    #[serde(rename = "type")]
+    pub type_flag: NumberType,
+}
 
-    Number {
-        value: f64,
-        raw: Atom,
-        #[serde(rename = "type")]
-        type_flag: NumberType,
-    },
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, EqIgnoreSpan)]
+#[cfg_attr(
+    feature = "rkyv",
+    derive(rkyv::Archive, rkyv::Serialize, rkyv::Deserialize)
+)]
+pub struct PercentageToken {
+    pub value: f64,
+    pub raw: Atom,
+}
 
-    Percentage {
-        value: f64,
-        raw: Atom,
-    },
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, EqIgnoreSpan)]
+#[cfg_attr(
+    feature = "rkyv",
+    derive(rkyv::Archive, rkyv::Serialize, rkyv::Deserialize)
+)]
+pub struct DimensionToken {
+    pub value: f64,
+    #[cfg_attr(feature = "rkyv", with(swc_atoms::EncodeJsWord))]
+    pub unit: JsWord,
+    #[serde(rename = "type")]
+    pub type_flag: NumberType,
+    /// Value and unit
+    pub raw: Box<(Atom, Atom)>,
+}
 
-    Dimension {
-        value: f64,
-        raw_value: Atom,
-        #[cfg_attr(feature = "rkyv", with(swc_atoms::EncodeJsWord))]
-        unit: JsWord,
-        raw_unit: Atom,
-        #[serde(rename = "type")]
-        type_flag: NumberType,
-    },
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, EqIgnoreSpan)]
+#[cfg_attr(
+    feature = "rkyv",
+    derive(rkyv::Archive, rkyv::Serialize, rkyv::Deserialize)
+)]
+pub struct WhiteSpaceToken {
+    pub value: Atom,
+}
 
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize, EqIgnoreSpan)]
+#[cfg_attr(
+    feature = "rkyv",
+    derive(rkyv::Archive, rkyv::Serialize, rkyv::Deserialize)
+)]
+#[cfg_attr(
+    feature = "rkyv",
+    archive(bound(
+        serialize = "__S: rkyv::ser::Serializer + rkyv::ser::ScratchSpace + \
+                     rkyv::ser::SharedSerializeRegistry",
+        deserialize = "__D: rkyv::de::SharedDeserializeRegistry"
+    ))
+)]
+pub enum Token {
```

Comment:
I think this is too much boxing.
Are you trying to optimive `memmove`?

Reply:
I tried different things, lexer is faster when we box tokens, because we use them in AST, i.e. `ComponentValues::PreservedTokens`, so memory usage was improved good, parser performance unfortunately was not very improved

Reply:
Also `Dimension`